File: mcspring-build/mcspring-plugin-manager/src/main/java/in/kyle/mcspring/manager/commands/CommandGamemode.java
package in.kyle.mcspring.manager.commands;

import org.bukkit.GameMode;
import org.bukkit.entity.Player;
import org.springframework.stereotype.Component;

import java.util.HashMap;
import java.util.Map;

import in.kyle.mcspring.command.Command;
import in.kyle.mcspring.subcommands.PluginCommand;

@Component
class CommandGamemode {
    
    @Command(value = "gamemode",
             aliases = "gm",
             description = "Set your game mode",
             usage = "/gamemode <creative|survival...>")
    void gamemode(PluginCommand command) {
        command.withPlayerSender("Only players can run this command.");
        Map<String, GameMode> gamemodes = new HashMap<>();
        for (GameMode value : GameMode.values()) {
            gamemodes.put(value.name().toLowerCase(), value);
            gamemodes.put(String.valueOf(value.getValue()), value);
        }
        command.withMap(gamemodes, s -> String.format("%s is not a valid game mode", s));
        command.then(this::gamemodeExecutor);
        
        command.otherwise("Usage: /gamemode <game mode>");
    }
    
    @Command(value = "gmc", description = "Set your game mode to creative")
    String gmc(Player sender) {
        return gamemodeExecutor(sender, GameMode.CREATIVE);
    }
    
    @Command(value = "gms", description = "Set your game mode to survival")
    String gms(Player sender) {
        return gamemodeExecutor(sender, GameMode.SURVIVAL);
    }
    
    String gamemodeExecutor(Player sender, GameMode gameMode) {
        sender.setGameMode(gameMode);
        return String.format("Game mode set to %s", gameMode.name().toLowerCase());
    }
}

File: mcspring-build/mcspring-plugin-manager/src/main/kotlin/in/kyle/mcspring/manager/commands/CommandGamemode.kt
package `in`.kyle.mcspring.manager.commands

import `in`.kyle.mcspring.commands.dsl.commandExecutor
import `in`.kyle.mcspring.commands.dsl.mcspring.Command
import org.bukkit.GameMode
import org.bukkit.entity.Player
import org.springframework.stereotype.Component
import kotlin.contracts.ExperimentalContracts

@ExperimentalContracts
@Component
internal class CommandGamemode {

    @Command(
            value = "gamemode",
            aliases = ["gm"],
            description = "Set your game mode",
            usage = "/gamemode <creative|survival...>"
    )
    fun gamemode() = commandExecutor {
        requirePlayer { message("Only players can run this command.") }

        val gameMode = mapArg<GameMode> {
            parser {
                map(GameMode.values().associateBy { it.name.toLowerCase() })
                map(GameMode.values().associateBy { it.value.toString() })
            }
            missing { message("Usage: /$label <game mode>") }
            invalid { message("Invalid game mode $it") }
        }

        then { message(gamemodeExecutor(sender as Player, gameMode)) }
    }

    private fun gamemodeExecutor(target: Player, gameMode: GameMode): String {
        target.gameMode = gameMode
        return "Game mode set to ${gameMode.name.toLowerCase()}"
    }

    @Command(value = "gmc", description = "Set your game mode to creative")
    fun gmc() = commandExecutor { then { message(gamemodeExecutor(sender as Player, GameMode.CREATIVE)) } }

    @Command(value = "gms", description = "Set your game mode to survival")
    fun gms() = commandExecutor { then { message(gamemodeExecutor(sender as Player, GameMode.SURVIVAL)) } }
}
